### ISSUE TITLE
Unable to install in ubuntu 14.04

### ISSUE THREAD

**Issue by akshay194** [bubenshchykov/ngrok#88]
I tried installing it in Ubuntu 14.04 but it does not work, please help.
```
$ sudo npm install ngrok -g
npm http GET https://registry.npmjs.org/ngrok
npm http 304 https://registry.npmjs.org/ngrok
npm http GET https://registry.npmjs.org/async
npm http GET https://registry.npmjs.org/decompress-zip
npm http GET https://registry.npmjs.org/lock
npm http GET https://registry.npmjs.org/request
npm http GET https://registry.npmjs.org/uuid
npm http 304 https://registry.npmjs.org/async
npm http 304 https://registry.npmjs.org/decompress-zip
npm http 304 https://registry.npmjs.org/lock
npm http 304 https://registry.npmjs.org/request
npm http 304 https://registry.npmjs.org/uuid
npm http GET https://registry.npmjs.org/types/node
npm http 404 https://registry.npmjs.org/types/node
npm ERR! TypeError: Cannot read property 'latest' of undefined
npm ERR!     at next (/usr/share/npm/lib/cache.js:687:35)
npm ERR!     at /usr/share/npm/lib/cache.js:675:5
npm ERR!     at saved (/usr/share/npm/node_modules/npm-registry-client/lib/get.js:142:7)
npm ERR!     at /usr/lib/nodejs/graceful-fs/polyfills.js:133:7
npm ERR!     at Object.oncomplete (fs.js:107:15)
npm ERR! If you need help, you may report this log at:
npm ERR!     <http://github.com/isaacs/npm/issues>
npm ERR! or email it to:
npm ERR!     <npm-@googlegroups.com>

npm ERR! System Linux 4.4.0-87-generic
npm ERR! command "/usr/bin/nodejs" "/usr/bin/npm" "install" "ngrok" "-g"
npm ERR! cwd /home/akshay
npm ERR! node -v v0.10.25
npm ERR! npm -v 1.3.10
npm ERR! type non_object_property_load
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/akshay/npm-debug.log
npm ERR! not ok code 0

```

**Comment by bubenshchykov** [comment]
boooo (
 404 https://registry.npmjs.org/types/node - can it be an npm glitch
Anyway, I updated ngrok on npm 20 seconds ago, can you try latest please

**Comment by bubenshchykov** [comment]
..if it doesn't help, try to update npm to latest ```npm i npm -g```
if that doesn't help neither, try to update node, if possible

**Comment by akshay194** [comment]
Tried after update, getting this:
```
$sudo npm install ngrok -g
/usr/bin/ngrok -> /usr/lib/node_modules/ngrok/bin/ngrok

> ngrok@2.2.17 postinstall /usr/lib/node_modules/ngrok
> node ./postinstall.js

ngrok - downloading binary https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip
events.js:182
      throw er; // Unhandled 'error' event
      ^

Error: EACCES: permission denied, open '/usr/lib/node_modules/ngrok/ngrok.zip'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! ngrok@2.2.17 postinstall: `node ./postinstall.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the ngrok@2.2.17 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/akshay/.npm/_logs/2017-08-08T10_58_16_164Z-debug.log
```
Node and npm version:
```
$ node -v
v8.2.1
$ npm -version
5.3.0

```

**Comment by bubenshchykov** [comment]
Yep, that's still unsolved :(
duplicate of https://github.com/bubenshchykov/ngrok/issues/87
can you try ```sudo chown -R $(whoami) $(npm config get prefix)/{lib/node_modules,bin,share}``` as npm suggests https://docs.npmjs.com/getting-started/fixing-npm-permissions

**Comment by bubenshchykov** [comment]
@gaoqiankun another report. do you have an idea what's wrong with permissions here?

**Comment by gaoqiankun** [comment]
I tried reproduce this issue, after many times failure, i finally reproduce it successfuly
```
❯ sudo npm install -g ngrok                                                                                                                                                                               
/usr/bin/ngrok -> /usr/lib/node_modules/ngrok/bin/ngrok
> ngrok@2.2.17 postinstall /usr/lib/node_modules/ngrok
> node ./postinstall.js

ngrok - downloading binary https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: EACCES: permission denied, open '/usr/lib/node_modules/ngrok/ngrok.zip'
    at Error (native)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! ngrok@2.2.17 postinstall: `node ./postinstall.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the ngrok@2.2.17 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/vagrant/.npm/_logs/2017-08-09T06_33_34_927Z-debug.log
```

I still can not undestand, why this happened, so I tried install an old version that before my code merged, then I got same problem, the only dirrerence is you catch the error, ant print it

```
[vagrant@postgres-mastrer] ~/work/ngrok (master) ⚡
❯ sudo npm install -g ngrok@2.2.11                                                                                                                                                                        ⏎
/usr/bin/ngrok -> /usr/lib/node_modules/ngrok/bin/ngrok

> ngrok@2.2.11 postinstall /usr/lib/node_modules/ngrok
> node ./postinstall.js

ngrok - downloading binary https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip ...
ngrok - error downloading binary. { Error: EACCES: permission denied, open '/usr/lib/node_modules/ngrok/bin/ngrok.zip'
    at Error (native)
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/usr/lib/node_modules/ngrok/bin/ngrok.zip' }
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! ngrok@2.2.11 postinstall: `node ./postinstall.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the ngrok@2.2.11 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/vagrant/.npm/_logs/2017-08-09T06_52_21_421Z-debug.log

```

I suppose that it is npm‘s `protecting mechanism`, so i viewed npm source code, and find it
```
// https://github.com/npm/npm/blob/latest/lib/utils/lifecycle.js#runCmd#line225
var user = unsafe ? null : npm.config.get('user')
var group = unsafe ? null : npm.config.get('group')
``` 

**npm will clear master process's `user` and `group`  if it think the subprocess is unsafe when it run lifecyle commands**

I will handler createstream permission error, and I found this little syntax bug will cause compatible problem
```
readStream
    .pipe(fs.createWriteStream(localFile))
    .on('finish', () => extract(cb));
```
It is my fault, i will fix it and create a pull request.

**Comment by bubenshchykov** [comment]
@gaoqiankun thx a lot for investigation!

**Comment by miquelbeltran** [comment]
Don't do `sudo chown -R $(whoami) $(npm config get prefix)/{lib/node_modules,bin,share}` as it might change the owner of everything inside `/usr/bin`, including `sudo` and you won't be able to repair the error without a full reinstall. This happens because `npm config get prefix` can be `/usr`

**Comment by bubenshchykov** [comment]
might be fixed in 2.2.19, @akshay194 can you try again?

**Comment by bubenshchykov** [comment]
should be fixed! feel free to re-open in needed
